---
title: "Index"
author: "Palma Luna Melissa"
date: "2025-02-02"
output:
  html_document:
    toc: true
    toc_depth: '5'
    df_print: paged
toc-title: "Contents"
---

# Differential expression analysis of the [recount3](https://doi.org/doi:10.18129/B9.bioc.recount3) project: **SRP124965**

A Differential expression analysis was conducted on the study **"Epigenome-wide Effects of Vitamin-D on Bronchial Epithelial Cells using ATAC-Seq and RNA-seq"**, obtained from the [Recount3 Study Explorer](https://jhubiostatistics.shinyapps.io/recount3-study-explorer/), to assess the effect of Vitamin D3 (also refereed as calcitriol) across different cell lines of human bronchial epithelial cells, utilizing various [Bioconductor](https://bioconductor.org) tools. 

**Project description in the Recount3:**
Vitamin D has been associated with viral respiratory infections, the main cause of severe asthma exacerbations in children. We used ATAC-Seq to evaluate the effect of vitamin D on chromatin accessibility in immortalized (BEAS-2B), normal (NHBEC), and asthma (AAEC) bronchial epithelial cells cultured and stimulated with calcitriol, poly I:C (to simulate viral infection), both, or sham (culture media), as well as the effect on gene expression. Overall design: ATAC-Seq and total RNA-Seq were performed on 4 asthma human bronchial epithelial cell samples (AAEC), 4 normal human bronchial epithelial cell samples (NHBEC) and 4 immortalized (BEAS-2B) samples, respectively treated with calcitriol, poly I:C, both or sham (culture media).

**Overview of the different treatments and cell lines for the experiment:**

Cell lines:

- Normal human bronchial epithelial
- Asthma human bronchial epithelial
- BEAS-2B immortalized human bronchial epithelial

Treatments:

- Vitamin D3 or calcitriol
- Poly (I:C)
- calcitriol + Poly (I:C)
- Sham (cultured media)

---

# Reduced Protocol

## Previous Steps
The fist part of the analysis consisted in the preparation of a [RangedSummarizedExperiment](https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) (RSE) object. The code for this step can be found in the R script named: `main_script.R`, for which fragments will be pasted below with additional comments relevant to the analysis: 

### Load Packages
```{R, message=FALSE, warning=FALSE}
## ---------------------------------
##            Path
## ---------------------------------
setwd("~/Desktop/Bioconductor_Final_project")
## ---------------------------------
##           R_packages
## ---------------------------------
library("recount3")
library("limma")
library("edgeR")
library("ggplot2")
library("pheatmap")
library("ExploreModelMatrix")
library("variancePartition")
```

### RSE object creation

```{R RSE object creation, message=FALSE, warning=FALSE}
## ---------------------------------
##         Data_Download
## ---------------------------------
# Download list of all human experiments
human_projects <- available_projects()

## Create a RangedSummarizedExperiment (RSE) object using the ID
## of the choosen project
project_vit_D <- create_rse(subset(human_projects,
  project == "SRP124965" & project_type == "data_sources"
))
```


### Explore and Modify the RSE object

```{R Explore_object, message=FALSE}
## ----------------------------------------
##             Explore_object
## ----------------------------------------

project_vit_D

```
The previous results revealed that converting the provided counts from base-pair count to read count is a necessary step.

```{R, message=FALSE, warning=FALSE}
## ------Change_to_read_counts------
assay(project_vit_D, "counts") <- compute_read_counts(project_vit_D)

##-------Facilitate Metadata access------
# project_vit_D$sra.sample_attributes[1:12] # inspect attributes
expand_sra_attributes(project_vit_D) ## No problems with the Data
## The metadata in the single the SRA (Sequence Read Archive) column is 
## separated into several ones in colData
project_vit_D <- expand_sra_attributes(project_vit_D)

## Showcases resultant columns 
colData(project_vit_D)[
  ,
  grepl("^sra_attribute", colnames(colData(project_vit_D)))
]
```
For the `sra_attribute columns` of interest, a change of type is needed, as they are categorical variables. Additionally a calculation is needed to determine the assigned gene proportion.

```{r}
## Change character to factor
project_vit_D$sra_attribute.cell_type <- factor(project_vit_D$sra_attribute.cell_type)
project_vit_D$sra_attribute.treatment <- factor(project_vit_D$sra_attribute.treatment)
project_vit_D$sra_attribute.source_name <- factor(project_vit_D$sra_attribute.source_name)

## Proportion of lectures assigned to genes
project_vit_D$assigned_gene_prop<-project_vit_D$recount_qc.gene_fc_count_all.assigned / project_vit_D$recount_qc.gene_fc_count_all.total
summary(project_vit_D$assigned_gene_prop)
## Note: The best sample has around 0.6430 of lectures assigned and 75% of the samples have less than 0.5697 assigned

## save data
unfiltered_vitamin_D<-project_vit_D

```


```{R}
##-------------------------------------------
##                  Filtering
##-------------------------------------------
hist(project_vit_D$assigned_gene_prop)
## In this step two options were plausible: Don't Filter samples and Filtering those below the Mean
## further analysis on both can be found at the no_filter.R and Below_Mean.R scripts

## get the mean
summary(unfiltered_vitamin_D$assigned_gene_prop) 
## Create object for this alternative analysis
below_MEAN_project_vit_D<-project_vit_D
```

As shown before in the histogram, the values are mainly between 0.3 and 0.65, with no indications of obvious extreme or outliers. Therefore two options for the filtering process arise:

- Only delete those genes with an expression below the 0.1
- Filter those samples below the Mean and then delete those genes with an expression below 0.1

The code for the fist option its available at the script `no_filter.R` and for the second option at the script `Below_Mean.R` 

Both scripts will be pasted below, with additional comments.

## Filtering

### Script 1: Don't Filter samples and Delete genes with a low expression

```{R no_filter 1, message=FALSE, warning=FALSE}
## ----------------------------------------
##        option 1: Don't Filter
## ----------------------------------------
## We are gonna continue with the Option: `Don't Filter samples`
## further analysis of the Option 2: `Filtering those below the Mean`
## is in the script Below_Mean.R

## Save the average expression of a gene in all samples
gene_means <- rowMeans(assay(project_vit_D, "counts"))
summary(gene_means)
```

Filter genes with a low expression, get the percent of remaining genes after filtering and the number of samples left.
```{R no_filter 2, message=FALSE, warning=FALSE}
## Delete genes with low mean expression
project_vit_D <- project_vit_D[gene_means > 0.1, ]

## Percent of remaining genes after filtering:
round(nrow(project_vit_D) / nrow(unfiltered_vitamin_D) * 100, 2)

# Number of samples:
dim(project_vit_D)
```

57.68 % of remaining genes and no samples deleted

### Script 2: Filter those samples below the Mean and Delete genes with a low expression

```{R message=FALSE, warning=FALSE}
## ----------------------------------------
## option 2: Filtering those below the Mean
## ----------------------------------------
## We are gonna continue with the Option: `Filtering those below the Mean`
## further analysis of the Option 1: `Filtering those below the Mean` is in the script no_filter.R
## Filter those below the mean (0.4891) using 0.4 instead of the complete number
below_MEAN_project_vit_D <- below_MEAN_project_vit_D[, below_MEAN_project_vit_D$assigned_gene_prop > 0.4]
## Save the average expression of a gene in all samples
gene_means_BM <- rowMeans(assay(below_MEAN_project_vit_D, "counts"))
```

Filter genes with a low expression, get the percent of remaining genes after filtering and the number of samples left.

```{R message=FALSE, warning=FALSE}
## Delete genes with low mean expression
below_MEAN_project_vit_D <- below_MEAN_project_vit_D[gene_means_BM > 0.1, ]
## Percent of remaining genes:
round(nrow(below_MEAN_project_vit_D) / nrow(unfiltered_vitamin_D) * 100, 2)
# Number of samples:
dim(below_MEAN_project_vit_D)
```

57.41%  of remaining genes and 3 samples deleted

## Filtering and Normalization
After filtering we continue with the normalization step, again both options separated in the scripts mentioned before.

### 1. Script 1: Don't Filter samples and Delete genes with a low expression

```{R no_filter 3, message=FALSE, warning=FALSE}
## ----------------------------------------
##          Data Normalization
## ----------------------------------------
## edgeR object creation
dge<-DGEList(
  counts = assay(project_vit_D, "counts"),
  genes = rowData(project_vit_D)
)
## Normalization with the TMM method (default)
dge <- calcNormFactors(dge)
```

### 2. Script 2: Filter those samples below the Mean and Delete genes with a low expression
```{R message=FALSE, warning=FALSE}
## ----------------------------------------
##          Data Normalization
## ----------------------------------------
## edgeR object creation
dge_BM<-DGEList(
  counts = assay(below_MEAN_project_vit_D, "counts"),
  genes = rowData(below_MEAN_project_vit_D)
)
## Normalization with the TMM method (default)
dge_BM <- calcNormFactors(dge_BM)
```


#### Boxplots

The boxplots below helps us visualize the distibution between the our samples, treatments and gene proportion. Visually, we identify the samples that were deleted in the previous filtering step in the Option 2.

```{R no_filter 4, message=FALSE, warning=FALSE}
## ----------------------------------------
##    Differential expression analysis
## ----------------------------------------
## ----------------------------
##            Graph
## ----------------------------
## -------------boxplot---------------
ggplot(as.data.frame(colData(project_vit_D)),
       aes(x = sra_attribute.treatment,y = assigned_gene_prop, color = sra_attribute.treatment)) +
  geom_boxplot() +
  facet_wrap(~ sra_attribute.cell_type) + # subgraph for each celltype
  theme_bw() +
  labs(title = "Op.1 Distribution of Assigned Gene Prop according to Treatment and Cell Type",
       x = "Treatment", y = "Assigned Gene Prop") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) # labels and text

```


```{R message=FALSE, warning=FALSE}
## ----------------------------------------
##    Differential expression analysis
## ----------------------------------------
## ----------------------------
##            Graph
## ----------------------------
## -------------boxplot---------------
ggplot(as.data.frame(colData(below_MEAN_project_vit_D)),
       aes(x = sra_attribute.treatment,y = assigned_gene_prop, color = sra_attribute.treatment)) +
  geom_boxplot() +
  facet_wrap(~ sra_attribute.cell_type) + # subgraph for each celltype
  theme_bw() +
  labs(title = "Op.2 Below Mean: Distribution of Assigned Gene Prop according to Treatment and Cell Type",
       x = "Treatment", y = "Assigned Gene Prop") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) # labels and text

```

## Differential expression analysis

Initially the chosen model was: 
`~ 0 + sra_attribute.cell_type + sra_attribute.treatment + sra_attribute.cell_type:sra_attribute.treatment + assigned_gene_prop`. However the following error arose:  
`Error in .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  : No residual degrees of freedom in linear model fits` After tracing back the error origin ,its appears that linear model lacks sufficient residual degrees of freedom to estimate the variability and perform the Bayes correction, since each combination has only one replicate.

The same error repeats in both cases worked on this analysis. 

### Correction of the objective

To assess the effect of Vitamin D3 across different cell lines of human bronchial epithelial cells, as stated in our objective, a necessary step is revaluate our model.

One option, is to use the following model `~ 0 + sra_attribute.cell_type + sra_attribute.treatment`, which assumes that the effects of Vitamin D3 are the same for all cell types. This model is useful to asses the effects of cell types and treatment on gene expression.

This model is used on the scripts no_filter and below_Mean. And its code and results will be shown:
For the first script: no_filter
```{R}
## ----------------------------
## Creating the statistic model
## ----------------------------
## Model as a Function of the variables treatment, cell_type and assigned_gene_prop
models <- model.matrix(~ 0 + sra_attribute.cell_type + sra_attribute.treatment + assigned_gene_prop,
                    data = colData(project_vit_D)
                    )
## names of the columns in the matrix
colnames(models)

## ----------------------------
##            Graph
## ----------------------------
## -----Dispersion graph of mean-variance trend----
vGene <- voom(dge, models, plot = TRUE)
## ----DE-----
## Fitting a linear Model and using an empirical Bayes Method
eb_results <- eBayes(lmFit(vGene))
## Extract a table of the DE genes
de_results <- topTable(
  eb_results,
  coef = 2,
  number = nrow(project_vit_D),
  sort.by = "none"
)
```
For the second script: Below_mean

```{R}
## ----------------------------
## Creating the statistic model
## ----------------------------
## Model as a Function of the variables treatment, cell_type and assigned_gene_prop
models_BM<- model.matrix(~ 0 + sra_attribute.cell_type + sra_attribute.treatment,
                           data = colData(below_MEAN_project_vit_D)
)
## names of the columns in the matrix
# colnames(models_BM)

## ----------------------------
##            Graph
## ----------------------------
## Dispersion graph of mean-variance trend
vGene_BM <- voom(dge_BM, models_BM, plot = TRUE)
## ----DE-----
## Fitting a linear Model and using an empirical Bayes Method
eb_results_BM <- eBayes(lmFit(vGene_BM))
## Extract a table of the DE genes
de_results_BM <- topTable(
  eb_results_BM,
  coef = 2,
  number = nrow(below_MEAN_project_vit_D),
  sort.by = "none"
)
```

Another option, and the one more suitable for this analysis is [variancePartition](https://bioconductor.org/packages/release/bioc/vignettes/variancePartition/inst/doc/variancePartition.html), which allows us consider the effects of different cell lines on gene expression using a linear mixed model, and is more suited for data with less replicates. 

For the data generated in the first script: no_filter:

```{R no_filter 5, include=FALSE}
## Load library
library("variancePartition")

## VariancePartition Model 
## Specify variables to consider
model <- ~ sra_attribute.treatment + (1 | sra_attribute.cell_type) + assigned_gene_prop
## Models treatment as the fixed effect and cell types as the random effects
## For each cell type, there is a baseline expression level that the model will adjust for
## ----------------------------
##            Graph
## ----------------------------
## -----Dispersion graph of mean-variance trend----
vGene <- voom(dge, plot = TRUE) 
fit <- dream(vGene, model, colData(project_vit_D))  # Model adjustment
print(fit)
fit <- eBayes(fit)
results <- topTable(fit, coef = "sra_attribute.treatment", number = Inf)
```


# Results
```{R}
## ----------------------------------------
##               Results
## ----------------------------------------
## ---------------------------
##            Graph
## ---------------------------
## Relationship between average and the change in gene expression
plotMA(eb_results, coef = 2)
## ----------------------------
##            Graph
## ---------------------------
## ----Volcano Plot----
## Statistical significance and magnitude of the change in expression
## highlight 3 of the most significant genes
volcanoplot(eb_results, coef = 2, highlight = 3, names = de_results$gene_name)
```

```{R}
# ----------------------------------------
##               Results
## ----------------------------------------
## ---------------------------
##            Graph
## ---------------------------
## Relationship between average and the change in gene expression
plotMA(eb_results_BM, coef = 2)
## ----------------------------
##            Graph
## ---------------------------
## ----Volcano Plot----
## Statistical significance and magnitude of the change in expression
## highlight 3 of the most significant genes
volcanoplot(eb_results_BM, coef = 2, highlight = 3, names = de_results_BM$gene_name)
```

## Heatmap and final graphics

Continuing this analysis with the 
```{R message=FALSE, warning=FALSE}
## ----------------------------------------
##          Visualizing DE genes
## ----------------------------------------
## ----------------------------
##            Graph
## ---------------------------
## ----Heatmap----
DE_no_filter_df <- as.data.frame(colData(project_vit_D)[, c("sra_attribute.cell_type", "sra_attribute.treatment", "assigned_gene_prop")])

colnames(DE_no_filter_df) <- c("cell_type", "Treatment", "gene_prop")
## get the 50 most significant genes
exprs_heatmap <- vGene$E[rank(de_results$adj.P.Val) <= 50, ]

identical(rowRanges(project_vit_D)$gene_id, de_results$gene_id)

## Change IDs to names
gene_names <- rownames(exprs_heatmap)
rownames(exprs_heatmap) <- rowRanges(project_vit_D)$gene_name[
  match(rownames(exprs_heatmap), rowRanges(project_vit_D)$gene_id)
]
## Create Heatmap
pheatmap(
  exprs_heatmap,
  cluster_rows = TRUE,
  cluster_cols = TRUE,
  show_rownames = TRUE,
  show_colnames = FALSE,
  annotation_col = DE_no_filter_df ## labels
)

```

```{R message=FALSE, warning=FALSE}
## ----------------------------------------
##          Visualizing DE genes
## ----------------------------------------
## ----------------------------
##            Graph
## ---------------------------
## ----Heatmap----
DE_below_MEan_df <- as.data.frame(colData(
  below_MEAN_project_vit_D)[, c("sra_attribute.cell_type", "sra_attribute.treatment", "assigned_gene_prop")])

colnames(DE_below_MEan_df) <- c("cell_type", "Treatment", "gene_prop")
## get the 50 most significant genes
BM_exprs_heatmap <- vGene_BM$E[rank(de_results_BM$adj.P.Val) <= 50, ]

# identical(rowRanges(below_MEAN_project_vit_D)$gene_id, de_results_BM$gene_id)

## Change IDs to names
gene_names_BM <- rownames(BM_exprs_heatmap)
rownames(BM_exprs_heatmap) <- rowRanges(below_MEAN_project_vit_D)$gene_name[
  match(rownames(BM_exprs_heatmap), rowRanges(below_MEAN_project_vit_D)$gene_id)
]
## Create Heatmap
pheatmap(
  BM_exprs_heatmap,
  cluster_rows = TRUE,
  cluster_cols = TRUE,
  show_rownames = TRUE,
  show_colnames = FALSE,
  annotation_col = DE_below_MEan_df ## labels
)
```

# Biological Conclusions
